lib/projfs.c

```diff
@@ -1156,6 +1157,126 @@ static int check_dir_empty(const char *path)
 	}
 }
 
+/**
+ * Uses the f*xattr family of functions to check that xattrs are supported on
+ * the supplied open file descriptor.
+ */
+static int test_xattr_support(int fd)
+{
+	int res;
+	char v = 1;
+
+	res = fsetxattr(fd, USER_PROJECTION_TEST, &v, 1, 0);
+	if (res == -1)
+		return -1;
+
+	res = fgetxattr(fd, USER_PROJECTION_TEST, &v, 1);
+	if (res != 1 || v != 1)
+		return -1;
+
+	res = fremovexattr(fd, USER_PROJECTION_TEST);
+	if (res != 0)
+		return -1;
+
+	return 0;
+}
+
+/**
+ * Sets the empty xattr on the given fd, and checks that it was set
+ * successfully. */
+static int set_and_check_empty_xattr(int fd)
+{
+	char v = 1;
+	int res;
+
+	res = fsetxattr(fd, USER_PROJECTION_EMPTY, &v, 1, 0);
+	if (res == -1)
+		return -1;
+
+	res = fgetxattr(fd, USER_PROJECTION_EMPTY, &v, 1);
+	if (res != 1 || v != 1)
+		return -1;
+
+	return 0;
+}
+
+#define SPARSE_TEST_FILENAME ".libprojfs-sparse-test"
+#define SPARSE_TEST_SIZE_BYTES 1048576
+
+/**
+ * Test that using ftruncate to create a sparse file works.
+ */
+static int test_sparse_support(struct projfs *fs)
+{
+	struct stat attrs;
+	int fd, res;
+
+	fd = openat(
+		fs->lowerdir_fd,
+		SPARSE_TEST_FILENAME,
+		O_CREAT | O_TRUNC | O_RDWR,
+		0600);
+	if (fd == -1) {
+		fprintf(stderr, "projfs: cannot openat sparse test file: %s\n",
+			strerror(errno));
+		return -1;
+	}
+
+	res = fstat(fd, &attrs);
+	if (res == -1) {
+		fprintf(stderr, "projfs: cannot fstat sparse test file: %s\n",
+			strerror(errno));
+		goto err_close;
+	}
+
+	if (attrs.st_size != 0) {
+		fprintf(stderr, "projfs: sparse test file should be 0 bytes, "
+				"is %ld bytes\n",
+			attrs.st_size);
+		goto err_close;
+	}
+
+	res = ftruncate(fd, SPARSE_TEST_SIZE_BYTES);
+	if (res == -1) {
+		fprintf(stderr, "projfs: cannot ftruncate sparse test "
+				"file: %s\n",
+			strerror(errno));
+		goto err_close;
+	}
+
+	res = fstat(fd, &attrs);
+	if (res == -1) {
+		fprintf(stderr, "projfs: cannot fstat sparse test file "
+				"after ftruncate: %s\n",
+			strerror(errno));
+		goto err_close;
+	}
+
+	if (attrs.st_size != SPARSE_TEST_SIZE_BYTES) {
+		fprintf(stderr, "projfs: sparse test file should be %d bytes, "
+				"is %ld bytes\n",
+			SPARSE_TEST_SIZE_BYTES,
+			attrs.st_size);
+		goto err_close;
+	}
+
+	/* check if the file consumes any actual blocks after ftruncate */
+	if (attrs.st_blocks > 0)
+		fprintf(stderr, "projfs: sparse files may not be "
+				"supported by the lower filesystem\n");
+
```

Comment:
Indeed, it is rather too much. (We can already pass `lowerdir_fd` only, we don't use anything else in `fs` as it is -- must've refactored something out.)

I had it this way because originally one of these steps was failing and it was difficult to isolate which. It's probably not helpful any more. 👍